File: rsfonts.c
#include <fcntl.h>
#include <spawn.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/snapshot.h>
#include <unistd.h>
extern char **environ;

int run_cmd(char *cmd)
{
    pid_t pid;
    char *argv[] = {"sh", "-c", cmd, NULL};
    int status = posix_spawn(&pid, "/bin/sh", NULL, NULL, argv, environ);
    if (status == 0) {
        if (waitpid(pid, &status, 0) == -1) {
            perror("waitpid");
        }
    }
    return status;
}

int main()
{
    if (geteuid() != 0) {
        printf("Run this as root!\n");
        exit(1);
    }
    
    int dirfd = open("/", O_RDONLY, 0);
    if (dirfd < 0) {
        perror("open");
        exit(1);
    }
    
    struct attrlist alist = { 0 };
    char abuf[2048];
    
    alist.commonattr = ATTR_BULK_REQUIRED;
    
    int count = fs_snapshot_list(dirfd, &alist, &abuf[0], sizeof (abuf), 0);
    if (count < 0) {
        perror("fs_snapshot_list");
        exit(1);
    } else if (count == 0) {
        perror("No snapshot founded");
        exit(1);
    }
    
    char *p = &abuf[0];
    char *field = p;
    uint32_t len = *(uint32_t *)field;
    field += sizeof (uint32_t);
    attribute_set_t attrs = *(attribute_set_t *)field;
    field += sizeof (attribute_set_t);
    attrreference_t ar = *(attrreference_t *)field;
    char *name = field + ar.attr_dataoffset;
    
    bool exisit = 0;
    if (access("/mnt2", 0)) {
        run_cmd("mkdir /mnt2");
    } else {
        exisit = 1;
    }
    
    char command[200];
    sprintf(command, "mount_apfs -s %s / /mnt2", name);
    if (run_cmd(command) == 0) {
        run_cmd("rm -rf /System/Library/Fonts/*");
        run_cmd("cp -a /mnt2/System/Library/Fonts/* /System/Library/Fonts");
    } else {
        perror("Mount failed");
        exit(1);
    }
    
    sprintf(command, "umount -f %s@/dev/disk0s1s1", name);
    run_cmd(command);
    
    if (!exisit) {
        run_cmd("rm -rf /mnt2");
    }
    
    run_cmd("rm -rf /private/var/mobile/Library/Caches/com.apple.UIStatusBar/*");
    run_cmd("rm -rf /private/var/mobile/Library/Caches/com.apple.keyboards/images/*");
    
    run_cmd("killall -9 SpringBoard");
    
    return 0;
}

File: rsfonts.m
#import <Foundation/Foundation.h>
#import <copyfile.h>
#import <removefile.h>
#import <sys/mount.h>
#import <sys/snapshot.h>
#import <sys/stat.h>

@interface LSApplicationProxy : NSObject<NSSecureCoding>

@property (nonatomic, readonly) NSURL *dataContainerURL;

+ (id)applicationProxyForIdentifier:(id)arg1;

@end

void removeCacheFromID(NSString *bundleID) {
    removefile([[[[[LSApplicationProxy applicationProxyForIdentifier:bundleID] dataContainerURL] path] stringByAppendingString:@"/Library/Caches/TelephonyUI-7"] UTF8String], NULL, REMOVEFILE_RECURSIVE);
    removefile([[[[[LSApplicationProxy applicationProxyForIdentifier:bundleID] dataContainerURL] path] stringByAppendingString:@"/Library/Caches/TelephonyUI-8"] UTF8String], NULL, REMOVEFILE_RECURSIVE);
}

int main() {
    if (getuid() != 0) {
        setuid(0);
    }

    if (getuid() != 0) {
        printf("Can't set uid as 0.\n");
        return 1;
    }

    clock_t start, stop;
    start = clock();
    printf("Fonts recovering...\n");

    int dirfd = open("/", O_RDONLY, 0);
    if (dirfd < 0) {
        perror("open");
        return 2;
    }

    struct attrlist alist = { 0 };
    char abuf[2048];

    alist.commonattr = ATTR_BULK_REQUIRED;

    int count = fs_snapshot_list(dirfd, &alist, &abuf[0], sizeof (abuf), 0);
    if (count < 0) {
        perror("fs_snapshot_list");
        return 3;
    } else if (count == 0) {
        printf("No snapshot founded.\n");
        return 4;
    }

    char *p = &abuf[0];
    char *field = p;
    field += sizeof (uint32_t);
    field += sizeof (attribute_set_t);
    attrreference_t ar = *(attrreference_t *)field;
    char *name = field + ar.attr_dataoffset;

    bool existed = false;

    if ([[NSFileManager defaultManager] fileExistsAtPath:@"/mnt2" isDirectory:&existed]) {
        if (!existed) {
            removefile("/mnt2", NULL, REMOVEFILE_RECURSIVE);
        }
    }
    if (!existed) {
        mkdir("/mnt2", 00755);
    }

    count = fs_snapshot_mount(dirfd, "/mnt2", name, 0);
    close(dirfd);
    if (count < 0) {
        perror("fs_snapshot_mount");
        return 5;
    }

    removefile("/System/Library/Fonts", NULL, REMOVEFILE_RECURSIVE);
    copyfile("/mnt2/System/Library/Fonts", "/System/Library/Fonts", NULL, COPYFILE_ALL | COPYFILE_RECURSIVE);

    unmount("/mnt2", MNT_FORCE);

    if (!existed) {
        removefile("/mnt2", NULL, REMOVEFILE_RECURSIVE);
    }

    removefile("/private/var/mobile/Library/Caches/com.apple.UIStatusBar", NULL, REMOVEFILE_RECURSIVE);
    removefile("/private/var/mobile/Library/Caches/com.apple.keyboards/images", NULL, REMOVEFILE_RECURSIVE);
    removefile("/private/var/mobile/Library/Caches/TelephonyUI-7", NULL, REMOVEFILE_RECURSIVE);
    removefile("/private/var/mobile/Library/Caches/TelephonyUI-8", NULL, REMOVEFILE_RECURSIVE);
    removeCacheFromID(@"com.apple.mobilephone");
    removeCacheFromID(@"com.apple.InCallService");
    removeCacheFromID(@"com.apple.CoreAuthUI");

    stop = clock();
    double duration = (double)(stop-start)/CLOCKS_PER_SEC;
    printf("Fonts recover succeeded.\nUsed %f seconds.\n", duration);

    return 0;
}
